### PR TITLE
Improve haddock comments

### PR DESCRIPTION
**Pull request by kokobd**
## Changes

- Try my best to preserve the original indentation of record fields
- Add haddock comments to data constructors
- Add me as a code owner to this plugin. I'll keep this plugin updated.

## Limitations

- This PR doesn't include support for ghc-9.2. It will be added in a later one.

## Motivation

Previously, when you execute the "Generate fields comments" action on the following record:

```haskell
data Record = Record
  { a :: Int
  , b :: String
  }
```

You will get:

```haskell
data Record = Record
  {
                -- | 
                a :: Int
  ,
                -- | 
                b :: String
  }
```

I think this is not satisfactory, and we should preserve the original indentation as much as we can.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3207"></a>

### CODE REVIEW

**Review by pepeiborra — APPROVED** [review]
This looks great, the new code is very nicely commented and structured. Thanks!

File: plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments/Data.hs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+emptyPriorHaddockComment :: Comment
+emptyPriorHaddockComment = mkComment "-- |"
+#if MIN_VERSION_ghc(9,0,0)
+    badRealSrcSpan
+#else
+    noSrcSpan
+#endif
+
+-- | Determines the given node has haddock comments attached to it.
+hasHaddock :: Data a => Anns -> Located a -> Maybe Bool
```
Comment (michaelpj) [inline plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments/Data.hs:147]: Should this just return `False` in the `Nothing` case? Or is it useful to be able to distinguish "don't know"?

**Review by michaelpj — APPROVED** [review]
Looks good and sensible. It seems like we go to some lengths to recognize Haddock - is this something we could ask upstream to tell us? i.e. identify comments as haddock comments when they are?

File: plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments/Data.hs (hunk 1 above)
Comment (kokobd) [reply plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments/Data.hs:147]: I think it's better to leave the caller to decide what to do when the given `Located a` is missing in the `Anns`. For example, the caller use `False` as the condition to enable the action for adding new haddock comments, but if it's `Nothing`, we should disable this action.

**Comment by kokobd** [comment]
> Looks good and sensible. It seems like we go to some lengths to recognize Haddock - is this something we could ask upstream to tell us? i.e. identify comments as haddock comments when they are?

@michaelpj Oh, I actually didn't expect this to be merged so quickly ;) I'll check that in another PR, when handling ghc 9.2 support. `ghc-exactprint` surely doesn't take care of this for us, but maybe `haddock` has some related library functions

**Pull request merged**